sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json

```diff
@@ -0,0 +1,1322 @@
+{
+  "spdxVersion": "SPDX-2.3",
+  "dataLicense": "CC0-1.0",
+  "SPDXID": "SPDXRef-DOCUMENT",
+  "creationInfo": {
+    "created": "2006-08-14T02:34:56-06:00",
+    "creators": [
+      "example SPDX document only"
+    ]
+  },
+  "name": "poppler-21.01.0-19.el9",
+  "packages": [
+    {
+      "SPDXID": "SPDXRef-SRPM",
+      "name": "poppler",
+      "versionInfo": "21.01.0-19.el9",
+      "supplier": "Organization: Red Hat",
+      "downloadLocation": "NOASSERTION",
+      "packageFileName": "poppler-21.01.0-19.el9.src.rpm",
+      "licenseConcluded": "(GPLv2 OR GPLv3) AND GPLv2+ AND LGPLv2+ AND MIT",
+      "externalRefs": [
+        {
+          "referenceCategory": "PACKAGE-MANAGER",
+          "referenceType": "purl",
+          "referenceLocator": "pkg:rpm/redhat/poppler@21.01.0-19.el9?arch=src"
+        }
+      ],
+      "checksums": [
+        {
+          "algorithm": "SHA256",
+          "checksumValue": "801331801a5d8209b5aaf0cc28f8cdc1f94a4bb349ef2f9a4300c4889bb06e3a"
+        }
+      ]
+    },
+    {
+      "SPDXID": "SPDXRef-Source0",
+      "name": "poppler",
+      "versionInfo": "21.01.0",
+      "downloadLocation": "http://poppler.freedesktop.org/poppler-21.01.0.tar.xz",
+      "packageFileName": "poppler-21.01.0.tar.xz",
+      "checksums": [
+        {
+          "algorithm": "SHA256",
+          "checksumValue": "016dde34e5f868ea98a32ca99b643325a9682281500942b7113f4ec88d20e2f3"
+        }
```

Comment:
Added these where a download URL is available.

Reply:
Actually, a question about this. For openssl we now include this:
```
    {
      "SPDXID": "SPDXRef-Source0",
      "name": "openssl",
      "versionInfo": "3.0.7",
      "downloadLocation": "https://openssl.org/source/openssl-3.0.7.tar.gz",
      "packageFileName": "openssl-3.0.7-hobbled.tar.gz",
      "checksums": [
        {
          "algorithm": "SHA256",
          "checksumValue": "4105046836812ed422922f851a57500118a99cc0f009b7eff2b3436110393377"
        }
      ],
      "externalRefs": [
        {
          "referenceCategory": "PACKAGE-MANAGER",
          "referenceType": "purl",
          "referenceLocator": "pkg:generic/openssl@3.0.7?download_url=https://openssl.org/source/openssl-3.0.7.tar.gz&checksum=sha256:4105046836812ed422922f851a57500118a99cc0f009b7eff2b3436110393377"
        }
      ]
    },
```

But of course the checksum is wrong in the purl. We don't have the original archive to checksum.

Technically it should probably be:

```
  "packages": [
    ...
    {
      "SPDXID": "SPDXRef-openssl-3.0.7",
      "name": "openssl",
      "versionInfo": "3.0.7",
      "downloadLocation": "https://openssl.org/source/openssl-3.0.7.tar.gz",
      "externalRefs": [
        {
          "referenceCategory": "PACKAGE-MANAGER",
          "referenceType": "purl",
          "referenceLocator": "pkg:generic/openssl@3.0.7?download_url=https://openssl.org/source/openssl-3.0.7.tar.gz"
        }
      ]
    },
    {
      "SPDXID": "SPDXRef-Source0",
      "name": "openssl",
      "versionInfo": "3.0.7",
      "packageFileName": "openssl-3.0.7-hobbled.tar.gz",
      "checksums": [
        {
          "algorithm": "SHA256",
          "checksumValue": "4105046836812ed422922f851a57500118a99cc0f009b7eff2b3436110393377"
        }
      ]
    },
  ...
  ],
  "relationships": [
    ...
    {
      "spdxElementId": "SPDXRef-Source0",
      "relationshipType": "GENERATED_FROM",
      "relatedSpdxElement": "SPDXRef-openssl-3.0.7"
    },
    ...
  ]
```

WDYT?

Reply:
Mmm, good catch. I wonder if we care about adding a separate node for `openssl-3.0.7-hobbled.tar.gz` then. It's the internally cached version of what goes into the SRPM that nobody can access really, but the SRPM contains that same content (essentially). Maybe the relationship should go straight from the openssl.org archive to the SRPM (via "contains")?

Reply:
Discussing this with tmlcoch I now think this would be handled with a "mid-stream" style upstream repository. I've re-factored the examples to mock that.

Reply:
But then we're losing the relationship to the original upstream component, no? Or is the expectation that the midstream repo would itself contain an SBOM that would document the relationship to its upstream counterpart?

Reply:
That's true. (Even if the midstream repo had such an SBOM, we'd also need a process to ingest _that_.)

For the use-case of mapping an upstream project version to all built artifacts containing that version, what do you think will embody "upstream project version"? Will it be package name and versionInfo, or will it be purl?

Reply:
> For the use-case of mapping an upstream project version to all built artifacts containing that version, what do you think will embody "upstream project version"? Will it be package name and versionInfo, or will it be purl?

Ideally both, but it's really the location of where it came from that is important. If tomorrow I decide to publish an `openssl-3.0.7.tar.gz` archive on https://foo.xyz, it will be the purl's `download_url` that sets differentiates it from the actual, official OpenSSL tar ball.

I can also imagine that once we have more of this data, you could build an allowlist of upstreams so that if an RPM includes sources from e.g. openssl.org in one build, it will be flagged as suspicious if the next time you try to build it it tries to pre-fetch sources from foo.xyz.

Reply:
I've pushed a new change which models openssl with a midstream repository that reports the origin URL and digest in a way that we can pull it into our SBOM.